intermediate_source/model_parallel_tutorial.py

```diff
@@ -0,0 +1,321 @@
+# -*- coding: utf-8 -*-
+"""
+Model Parallel Best Practices
+*************************************************************
+**Author**: `Shen Li <https://mrshenli.github.io/>`_
+
+Data parallel and model parallel are widely-used distributed training
+techniques. Previous posts have explained how to use
+`DataParallel <https://pytorch.org/tutorials/beginner/blitz/data_parallel_tutorial.html>`_
+to train a neural network on multiple GPUs. ``DataParallel`` replicates the same
+model to all GPUs, where each GPU consumes a different partition of the input
+data. Although it can significantly accelerate the training process, it does not
+work for some use cases where the model is large to fit into a single GPU. This
+post shows how to solve that problem by using model parallel and also shares
+some insights on how to speed up model parallel training.
+
+The high-level idea of model parallel is to place different sub-networks of a
+model onto different devices. All input data will run through all devices, but
+each device only operates on a part of the entire model. In this post, we will
+not try to construct huge models and squeeze them into a limited number of GPUs.
+Instead, this post focuses on showing the idea of model parallel. It is up to
+the readers to apply the ideas to real-world applications.
+
+Let us start with a toy model that contains two linear layers. To run this
+model on two GPUs, simply put each linear layer on a different GPU, and move
+inputs and intermediate outputs to match the layer devices accordingly.
+"""
+
+import torch
+import torch.nn as nn
+import torch.optim as optim
+
+
+class ToyModel(nn.Module):
+    def __init__(self):
+        super(ToyModel, self).__init__()
+        self.net1 = torch.nn.Linear(10, 10).cuda(0)
+        self.net2 = torch.nn.Linear(10, 5).cuda(1)
+
+    def forward(self, x):
+        return self.net2(self.net1(x.cuda(0)).cuda(1))
+
+######################################################################
+# Note that, the above ``ToyModel`` looks very similar to how one would
+# implement it on a single GPU, except the four ``cuda(device)`` calls which
+# place linear layers and tensors to on proper devices. That is the only
+# place in the model that requires changes. The ``backward()`` and
+# ``torch.optim`` will automatically take care of gradients as if the
+# model is on one GPU. You only need to make sure that the labels are on the
+# same device as the outputs when calling the loss function.
```

Comment:
is it possible to highlight the changes?

Reply:
@brianjo is there a way to highlight a line in the code? Thanks!